qiskit/algorithms/gradients/lin_comb_sampler_gradient.py

```diff
@@ -47,80 +71,69 @@ def __init__(self, sampler: BaseSampler, options: Options | None = None):
-                    result_idx += 1
-                    for grad_data in gradient_circuits_[param]:
-                        coeff = grad_data.coeff
-                        # if the parameter is a parameter expression, we need to substitute
-                        if isinstance(coeff, ParameterExpression):
-                            local_map = {
-                                p: parameter_values_[circuit.parameters.data.index(p)]
-                                for p in coeff.parameters
-                            }
-                            bound_coeff = float(coeff.bind(local_map))
-                        else:
-                            bound_coeff = coeff
-                        coeffs.append(bound_coeff)
-
+        jobs, metadata = [], []
+        for circuit, parameter_values_, parameter_set in zip(
+            circuits, parameter_values, parameter_sets
+        ):
+            # Prepare circuits for the gradient of the specified parameters.
+            meta = {"parameters": [p for p in circuit.parameters if p in parameter_set]}
+            metadata.append(meta)
+            circuit_key = _circuit_key(circuit)
+            if circuit_key not in self._lin_comb_cache:
+                self._lin_comb_cache[circuit_key] = _make_lin_comb_gradient_circuit(
+                    circuit, add_measurement=True
+                )
+            lin_comb_circuits = self._lin_comb_cache[circuit_key]
+            gradient_circuits = []
+            for param in circuit.parameters:
+                if param not in parameter_set:
+                    continue
+                gradient_circuits.append(lin_comb_circuits[param])
             n = len(gradient_circuits)
+
             job = self._sampler.run(gradient_circuits, [parameter_values_] * n, **options)
             jobs.append(job)
-            result_indices_all.append(result_indices)
-            coeffs_all.append(coeffs)
 
-        # combine the results
         try:
             results = [job.result() for job in jobs]
         except Exception as exc:
             raise AlgorithmError("Sampler job failed.") from exc
-
+        # Compute the gradients.
         gradients = []
         for i, result in enumerate(results):
             n = 2 ** circuits[i].num_qubits
-            grad_dists = np.zeros((len(metadata_[i]["parameters"]), n))
-            for idx, coeff, dist in zip(result_indices_all[i], coeffs_all[i], result.quasi_dists):
-                grad_dists[idx][list(dist.keys())[:n]] += np.array(list(dist.values())[:n]) * coeff
-                grad_dists[idx][list(dist.keys())[:n]] -= np.array(list(dist.values())[n:]) * coeff
-
-            gradient_ = []
-            for grad_dist in grad_dists:
-                gradient_.append(dict(enumerate(grad_dist)))
-            gradients.append(gradient_)
+            gradient = []
+            for dist in result.quasi_dists:
+                grad_dist = np.zeros(2 ** circuits[i].num_qubits)
```

Comment:
Also here it would be good to avoid the exponentially large distribution per default 🙂